File: src/main/java/com/refinedmods/refinedstorage/screen/FluidAmountScreen.java
package com.refinedmods.refinedstorage.screen;

import com.refinedmods.refinedstorage.RS;
import com.refinedmods.refinedstorage.container.FluidAmountContainerMenu;
import com.refinedmods.refinedstorage.network.SetFluidFilterSlotMessage;
import com.refinedmods.refinedstorage.util.StackUtils;
import net.minecraft.client.gui.screens.Screen;
import net.minecraft.network.chat.Component;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.entity.player.Player;
import net.minecraftforge.fluids.FluidStack;
import org.apache.commons.lang3.tuple.Pair;

import javax.annotation.Nullable;
import java.util.function.UnaryOperator;

public class FluidAmountScreen extends AmountSpecifyingScreen<FluidAmountContainerMenu> {
    private static final ResourceLocation TEXTURE_WIDE = new ResourceLocation(RS.ID, "textures/gui/amount_specifying_wide.png");
    private static final ResourceLocation TEXTURE = new ResourceLocation(RS.ID, "textures/gui/amount_specifying.png");

    private final int containerSlot;
    private final FluidStack stack;
    private final int maxAmount;
    @Nullable
    private final UnaryOperator<Screen> alternativesScreenFactory;

    public FluidAmountScreen(BaseScreen parent, Player player, int containerSlot, FluidStack stack, int maxAmount, @Nullable UnaryOperator<Screen> alternativesScreenFactory) {
        super(parent, new FluidAmountContainerMenu(player, stack), alternativesScreenFactory != null ? 194 : 172, 99, player.getInventory(), Component.translatable("gui.refinedstorage.fluid_amount"));

        this.containerSlot = containerSlot;
        this.stack = stack;
        this.maxAmount = maxAmount;
        this.alternativesScreenFactory = alternativesScreenFactory;
    }

    @Override
    protected int getOkCancelButtonWidth() {
        return alternativesScreenFactory != null ? 75 : super.getOkCancelButtonWidth();
    }

    @Override
    public void onPostInit(int x, int y) {
        super.onPostInit(x, y);

        if (alternativesScreenFactory != null) {
            addButton(x + 114, cancelButton.getY() + 24, getOkCancelButtonWidth(), 20, Component.translatable("gui.refinedstorage.alternatives"), true, true, btn -> minecraft.setScreen(alternativesScreenFactory.apply(this)));
        }
    }

    @Override
    protected Pair<Integer, Integer> getOkCancelPos() {
        if (alternativesScreenFactory == null) {
            return super.getOkCancelPos();
        }

        return Pair.of(114, 22);
    }

    @Override
    protected int getDefaultAmount() {
        return stack.getAmount();
    }

    @Override
    protected boolean canAmountGoNegative() {
        return false;
    }

    @Override
    protected int getMaxAmount() {
        return maxAmount;
    }

    @Override
    protected Component getOkButtonText() {
        return Component.translatable("misc.refinedstorage.set");
    }

    @Override
    protected ResourceLocation getTexture() {
        return alternativesScreenFactory != null ? TEXTURE_WIDE : TEXTURE;
    }

    @Override
    protected int[] getIncrements() {
        return new int[]{
            100, 500, 1000,
            -100, -500, -1000
        };
    }

    @Override
    protected void onOkButtonPressed(boolean shiftDown) {
        try {
            int amount = Integer.parseInt(amountField.getValue());

            RS.NETWORK_HANDLER.sendToServer(new SetFluidFilterSlotMessage(containerSlot, StackUtils.copy(stack, amount)));

            close();
        } catch (NumberFormatException e) {
            // NO OP
        }
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/screen/FluidAmountScreen.kt
package com.refinedmods.refinedstorage.screen

import com.refinedmods.refinedstorage.RS
import com.refinedmods.refinedstorage.container.FluidAmountContainer
import com.refinedmods.refinedstorage.network.SetFluidFilterSlotMessage
import com.refinedmods.refinedstorage.util.StackUtils.copy
import net.minecraft.client.gui.screen.Screen
import net.minecraft.client.resources.I18n
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.util.text.Text
import net.minecraft.util.text.TranslationTextComponent
import net.minecraftforge.fluids.FluidInstance
import org.apache.commons.lang3.tuple.Pair
import java.util.function.Function

class FluidAmountScreen(parent: BaseScreen<*>?, player: PlayerEntity, private val containerSlot: Int, stack: FluidInstance?, maxAmount: Int, @Nullable alternativesScreenFactory: Function<Screen, Screen>?) : AmountSpecifyingScreen<FluidAmountContainer?>(parent, FluidAmountContainer(player, stack), if (alternativesScreenFactory != null) 194 else 172, 99, player.inventory, TranslationTextComponent("gui.refinedstorage.fluid_amount")) {
    private val stack: FluidInstance?
    private override val maxAmount: Int

    @Nullable
    private val alternativesScreenFactory: Function<Screen, Screen>?
    override fun getOkCancelButtonWidth(): Int {
        return if (alternativesScreenFactory != null) 75 else super.getOkCancelButtonWidth()
    }

    override fun onPostInit(x: Int, y: Int) {
        super.onPostInit(x, y)
        if (alternativesScreenFactory != null) {
            addButton(x + 114, cancelButton.y + 24, okCancelButtonWidth, 20, TranslationTextComponent("gui.refinedstorage.alternatives"), true, true, Button.IPressable({ btn -> minecraft.displayGuiScreen(alternativesScreenFactory.apply(this)) }))
        }
    }

    override fun getOkCancelPos(): Pair<Int?, Int?>? {
        return if (alternativesScreenFactory == null) {
            super.getOkCancelPos()
        } else Pair.of(114, 22)
    }

    override fun getDefaultAmount(): Int {
        return stack.getAmount()
    }

    override fun canAmountGoNegative(): Boolean {
        return false
    }

    override fun getMaxAmount(): Int {
        return maxAmount
    }

    override fun getOkButtonText(): Text {
        return TranslationTextComponent("misc.refinedstorage.set")
    }

    override fun getTexture(): String {
        return if (alternativesScreenFactory != null) "gui/amount_specifying_wide.png" else "gui/amount_specifying.png"
    }

    override fun getIncrements(): IntArray {
        return intArrayOf(
                100, 500, 1000,
                -100, -500, -1000
        )
    }

    override fun onOkButtonPressed(shiftDown: Boolean) {
        try {
            val amount = amountField!!.text.toInt()
            RS.NETWORK_HANDLER.sendToServer(SetFluidFilterSlotMessage(containerSlot, copy(stack, amount)))
            close()
        } catch (e: NumberFormatException) {
            // NO OP
        }
    }

    init {
        this.stack = stack
        this.maxAmount = maxAmount
        this.alternativesScreenFactory = alternativesScreenFactory
    }
}